src/sync/writer.rs

```diff
@@ -1080,29 +1080,34 @@ mod tests {
         assert_batches_eq!(expected, &results);
     }
 
+    type OldNewVal = (Vec<i32>, Vec<i32>, Option<Vec<&'static str>>);
+
     #[rstest]
-    #[case(vec![(vec![1, 2], vec![1, 2])])]
-    #[case(vec![(vec![1, 2], vec![2, 1]), (vec![1, 2], vec![2, 1])])]
+    #[case(vec![(vec![1, 2], vec![1, 2], None)])]
+    #[should_panic(expected = "assertion `left == right` failed")]
+    #[case(vec![
+        (vec![1, 2], vec![2, 1], Some(vec!["a", "b"])),
+        (vec![2, 1], vec![1, 2], Some(vec!["a", "b"]))
+    ])]
     #[case(vec![
```

Comment:
Turns out this is an edge case i missed; what happens is that physical squashing compresses
```rust
vec![1, 2], vec![2, 1], Some(vec!["a", "b"])
```
into
```rust
1, 1, b
```
and it compresses 
```rust
vec![2, 1], vec![1, 2], Some(vec!["a", "b"])
```
into 
```rust
2, 2, b
```
since it detects a PK-chain.

Consequently, the post-apply state ends up being
```rust
1, b
2, b
```
which is wrong.

This necessitates the we normalize the batch prior to physical squashing, or ideally just alter the protocol to how it was before (as it is resiliant against this edge case). Going back to the previous protocol version will also simplify a lot of the planning logic too.

Reply:
Actually, thinking about this some more, this seems to be an impossible input batch for the writer.

This is because it violates the PK-conflict rules, which is an invariant that I'm working with (i.e. upstream is responsible for ensuring the original PK rules are enforced and reflected in the change batches).

So what needs to be done is for the test case to get updated along those lines.

Reply:
Would this be fine if the test case was 2->3, 1->2, 3->1 (i.e. we swap the PKs using a third temporary one)? What if the user temporarily suspends the PK constraint during a transaction (or it's deferred to the end of the TX) https://www.postgresql.org/docs/current/sql-set-constraints.html and is still valid when the TX commits)?

Reply:
> Would this be fine if the test case was 2->3, 1->2, 3->1 (i.e. we swap the PKs using a third temporary one)?

Yes, the replacement test case does this https://github.com/splitgraph/seafowl/blob/b0efbd3306d30b96c3dd97ad40d606f27adb92a4/src/sync/writer.rs#L1098-L1101

> What if the user temporarily suspends the PK constraint during a transaction (or it's deferred to the end of the TX) https://www.postgresql.org/docs/current/sql-set-constraints.html and is still valid when the TX commits)?

We're vulnerable against that until we switch over to the upsert batch format.